### PR TITLE
Fixed metadata injection for local actions

### PR DESCRIPTION
**Pull request by mkrjn99**
> [!IMPORTANT]
> Adds metadata injection support for local actions in `ComposioToolSet` and tests it with a new test case.
> 
>   - **Behavior**:
>     - Adds `add_local_action_metadata()` in `ProcessorHelper` to handle metadata for local actions.
>     - Updates `execute_action()` in `ComposioToolSet` to use `add_local_action_metadata()` for local actions.
>   - **Testing**:
>     - Adds `test_metadata_injection.py` to verify metadata injection for local actions using `ComposioToolSet`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=ComposioHQ%2Fcomposio&utm_source=github&utm_medium=referral)<sup> for 29e6856500b00e3ece920a738d9e8f2c3fb196e0. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: python/composio/tools/toolset.py [hunk 1]
```diff
@@ -396,6 +396,11 @@ def add_metadata(self, action: Action, metadata: t.Optional[t.Dict]) -> t.Dict:
         metadata.update(self._get_metadata(key=action))
         return metadata
 
+    def add_local_action_metadata(self, action_type: ActionType, metadata: t.Optional[t.Dict]) -> t.Dict:
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py:399]: Consider adding a docstring to the `add_local_action_metadata` method to document its purpose, parameters and return value. This would help maintain consistency with the codebase's documentation standards. Example:

```python
def add_local_action_metadata(self, action_type: ActionType, metadata: t.Optional[t.Dict]) -> t.Dict:
    """Add metadata for local actions.
    
    Args:
        action_type (ActionType): The type of action being executed
        metadata (Optional[Dict]): Optional metadata dictionary to update
        
    Returns:
        Dict: Updated metadata dictionary with local action metadata
    """
```

File: python/tests/test_tools/test_metadata_injection.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+
+class RequestClass(BaseModel):
+    pass
+
+
+class ResponseClass(BaseModel):
+    pass
+
+metadata_value = None
```
Comment (shreysingla11) [inline python/tests/test_tools/test_metadata_injection.py:13]: Consider using a more descriptive name for the global variable and adding type hints:

```python
test_captured_metadata: Optional[str] = None  # Used to capture metadata in test scenarios
```

This makes the test code more maintainable and its purpose clearer.

File: python/tests/test_tools/test_metadata_injection.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+def action_metadata_injection(request: RequestClass, metadata: dict) -> ResponseClass:
+    """
+    Test metadata injection.
+    """
+    global metadata_value
+    metadata_value = metadata[API_KEY_KEY]
+    return ResponseClass()
+
+
+def test_metadata_injection() -> None:
```
Comment (shreysingla11) [inline python/tests/test_tools/test_metadata_injection.py:27]: Consider adding more test cases to cover different scenarios:

1. Test with None metadata
2. Test with existing metadata to ensure proper merging
3. Test with different action types

This would improve test coverage and ensure robustness of the metadata injection functionality.

**Comment by shreysingla11** [comment]
### Code Review Summary

Overall, the changes look well-structured and properly implement metadata injection for local actions. Here's a breakdown of the review:

#### Strengths:
- Clean implementation following existing patterns
- Good separation of concerns with dedicated method for local action metadata
- Basic test coverage with specific test case
- Type safety maintained with proper type hints

#### Suggestions for Improvement:
1. Documentation:
   - Add docstring to `add_local_action_metadata` method
   - Consider adding inline comments explaining the metadata injection flow

2. Testing:
   - Add more test cases covering edge cases and different scenarios
   - Improve test variable naming for clarity
   - Add type hints to test variables

3. Code Structure:
   - Consider adding validation for action_type parameter
   - Consider adding debug logging for metadata injection process

The changes are good to merge after addressing the documentation and test improvements. The core functionality appears solid and well-implemented.
